traits/observers/_trait_event_notifier.py

```diff
@@ -0,0 +1,207 @@
+
+    An ``TraitEventNotifier`` is a callable to be contributed to an
+    "observable" object. The call signature is defined by the observable object
+    and may vary. It is the responsibility of the ``event_factory`` to adapt
+    the varying call signatures and create an event object to be given
+    to the user's handler.
+
+    A ``TraitEventNotifier`` keeps a reference count in order to address
+    situations where a same object is repeated inside a container
+    and one would not want to fire the same change handler multiple times
+    (see enthought/traits#237). For that purpose, a notifier keeps track of
+    the ``HasTraits`` instance (called ``target``) on which the user applies
+    the observers, keeps a reference count internally, and it also needs to
+    determine whether another notifier refers to the same change handler and
+    ``HasTraits`` instance.
+
+    Since there is only one reference count associated with a notifier,
+    each notifier is expected to be added to only one observable.
+    """
+
+    def __init__(
+            self, *, handler, target,
+            event_factory, prevent_event, dispatcher):
+        """
+        Parameters
+        ----------
+        handler : callable(event)
+            The user's handler to receive the change event.
+            The event object is returned by the ``event_factory``.
+            If the handler is an instance method, then a weak reference is
+            created for the method. If the instance is garbage collected,
+            the notifier will be muted.
+        target : object
+            An object for defining the context of the notifier.
+            A weak reference is created for the target.
+            If the target is garbage collected, the notifier will be muted.
+            This target is typically an instance of ``HasTraits`` and will be
+            seen by the user as the "owner" of the change handler.
+            This is also used for distinguishing one notifier from another
+            notifier wrapping the same handler.
+        event_factory : callable(*args, **kwargs) -> object
+            A factory function for creating the event object to be sent to
+            the handler. The call signature must be compatible with the
+            call signature expected by the observable this notifier is used
+            with. e.g. for CTrait, the call signature will be
+            ``(object, name, old, new)``.
+        prevent_event : callable(event) -> boolean
+            A callable for controlling whether the user handler should be
+            invoked. It receives the event created by the event factory and
+            returns true if the event should be prevented, false if the event
+            should be fired.
+        dispatcher : callable(handler, event)
+            A callable for dispatching the handler, e.g. on a different
+            thread or on a GUI event loop. ``event`` is the object
+            created by the event factory.
```

Comment:
I am a bit tempted to group `handler, target, dispatcher` into another object. They do always go together in other places. That object can provide the methods for `equals` and for whether or not the weakrefs are still live. 

We could also refactor this later when all the other pieces are here.